### ISSUE TITLE
Expose block time

### ISSUE THREAD

**Issue by agraebe** [blockstack/stacks-blockchain#1777]
For the Explorer to show an estimate for the time it takes to complete a tx, we should expose the block time

**Comment by zone117x** [comment]
@kantai @jcnelson thoughts?

**Comment by kantai** [comment]
The core node itself isn't aware of the target block time -- it's just polling the Bitcoin node. It could return the amount of time it waits between blocks before commitments, but that doesn't really convey the same thing.

You could estimate the average block time by just sampling the block times that you receive in the explorer.

**Comment by zone117x** [comment]
Is average block time expected to deviate significantly from bitcoin's target block time? If not seems like it should just be hardcoded:
`mainnet=10minutes`, `testnet=2minutes`, or w/e

**Comment by agraebe** [comment]
friendly ping @kantai and @jcnelson :)

**Comment by kantai** [comment]
I don't think that the stacks-node should be hardcoded and displaying this information -- the stacks-node itself is not responsible for the timing of Bitcoin blocks. If this is useful for the explorer or sidecar, the values should be hardcoded there.

**Comment by diwakergupta** [comment]
Yep, agree w/ @kantai. I'm going to close this out, @agraebe please re-open if you disagree.

**Comment by Santy9983** [comment]
Hi, I was looking at the information about blockchain without anything specific for now, but considering that I am very interested in maintaining links for smart contracts of various projects.

Enviado desde mi iPhone

**Comment by zone117x** [comment]
@kantai are Stacks blocks intended to be at consistent pace with bitcoin blocks, assuming there's at least 1 miner?

**Comment by kantai** [comment]
> @kantai are Stacks blocks intended to be at consistent pace with bitcoin blocks, assuming there's at least 1 miner?

Yes, there should be a Stacks block produced in every bitcoin block as long as there's 1 miner.

**Comment by zone117x** [comment]
Thanks @kantai.
What is the current testnet's bitcoin block time? Is there a config file somewhere I can reference?

**Comment by psq** [comment]
@zone117x the testnet advances blocks every 30 seconds, but this is not part of the regular stacks-node configuration.  One example configuration is here I think https://github.com/blockstack/stacks-blockchain/blob/master/testnet/bitcoin-neon-controller/local-leader.toml.default#L3

On BTC mainnet, blocks will come in on __average__ every 10 minutes, or if they come in faster, or slower, the difficulty will get adjusted.  However, blocks could be generated very close to each other, making it very possible there will be no miners if this happens.

You can get a sense of how much the block time varies from https://btc.com/block.  For example, there was 49 seconds between 643,419 and 643,420, over 45 minutes between 643,400 and 643,401, and 12 seconds between 643,137 and 643,13, which makes the "average" not so useful as the distribution can be quite broad (I only looked at 2 days worth of transactions).

If a Stacks miner attempts to mine blocks that comes too close to each others, this is likely to create a fork, and one of the 2 blocks will get eventually discarded.

Furthermore, once microblock mining is enabled, this will further complicates things.

**Comment by zone117x** [comment]
Thanks @psq!
The thing I'm unsure about is the toml `block_time` vs the bitcoin node's block time. 

AFAIK the toml `block_time` was for mocknet (simulated burnchain). While the current testnet is running against an actual bitcoin node in regtest mode with presumably it's own configured block time?

**Comment by CharlieC3** [comment]
@zone117x The `block_time` for our Argon bitcoin node is currently `30000`, [you can find it here.](https://github.com/blockstackpbc/k8s/blob/master/blockstack-core/v2/argon/bitcoin/staging/configmap.yaml#L7)

**Comment by zone117x** [comment]
@CharlieC3 @kantai could I get a brief description of how the stacks-node block time relates to the bitcoin block time? I don't understand why there are both.

**Comment by kantai** [comment]
That `toml` definition for the block time isn't the `stacks-node`, it's a config option for the "bitcoin controller" for the regtest Bitcoin node. A regtest Bitcoin node won't mine a block automatically, so the controller just calls the mine block RPC.

**Comment by zone117x** [comment]
Ohhh that makes much more sense now. Thanks @kantai!

**Comment by zone117x** [comment]
We decided to have the sidecar return the current network time as a hardcoded payload, which clients can query as the source of truth. 

So will be set to 30 seconds right now, then 2 minutes after next reset (it is being bumped), and eventually 10 minutes once we match Bitcoin.
